libvast/builtins/operators/pseudonymize.cpp

```diff
@@ -0,0 +1,191 @@
+struct configuration {
+  // field for future extensibility; currently we only use the Crypto-PAn method
+  std::string method;
+  std::string seed;
+  std::array<address::byte_type, vast::address::pseudonymization_seed_array_size>
+    seed_bytes{};
+  std::vector<std::string> fields;
+
+  template <class Inspector>
+  friend auto inspect(Inspector& f, configuration& x) {
+    return f(x.method, x.seed, x.fields);
+  }
+
+  static inline const record_type& layout() noexcept {
+    static auto result = record_type{
+      {"method", string_type{}},
+      {"seed", string_type{}},
+      {"fields", list_type{string_type{}}},
+    };
+    return result;
+  }
+};
+
+class pseudonymize_operator : public pipeline_operator {
+public:
+  pseudonymize_operator(configuration config) : config_{std::move(config)} {
+    parse_seed_string();
+  }
+
+  /// Applies the transformation to an Arrow Record Batch with a corresponding
+  /// VAST layout.
+  [[nodiscard]] caf::error
+  add(type layout, std::shared_ptr<arrow::RecordBatch> batch) override {
+    std::vector<indexed_transformation> transformations;
+    auto transformation = [&](struct record_type::field field,
+                              std::shared_ptr<arrow::Array> array) noexcept
+      -> std::vector<
+        std::pair<struct record_type::field, std::shared_ptr<arrow::Array>>> {
+      auto builder
+        = address_type::make_arrow_builder(arrow::default_memory_pool());
+      auto address_view_generator = values(
+        address_type{}, caf::get<type_to_arrow_array_t<address_type>>(*array));
+      for (const auto& address : address_view_generator) {
+        auto append_status = arrow::Status{};
+        if (address) {
+          auto pseudonymized_address
+            = vast::address::pseudonymize(*address, config_.seed_bytes);
+          append_status
+            = append_builder(address_type{}, *builder, pseudonymized_address);
+        } else {
+          append_status = builder->AppendNull();
+        }
+        VAST_ASSERT(append_status.ok(), append_status.ToString().c_str());
+      }
+      auto new_array = builder->Finish().ValueOrDie();
+      return {
+        {field, new_array},
+      };
+    };
+    for (const auto& field_name : config_.fields) {
+      for (const auto& index : caf::get<record_type>(layout).resolve_key_suffix(
+             field_name, layout.name())) {
+        if (!caf::holds_alternative<address_type>(
+              caf::get<record_type>(layout).field(index).type)) {
+          VAST_ASSERT(false, "record batch field to be pseudonymized but does "
+                             "not "
+                             "have address type");
+          VAST_WARN(fmt::format("Field {} is to be pseudonymized but does not "
+                                "contain "
+                                "IP "
+                                "address values; skipping pseudonymization",
+                                field_name));
```

Comment:
I think this can just be a debug message; it's not worth of a warning, and it's certainly not something that should trigger an assertion.

The `VAST_*` family of macros also already do the formatting for you, so there's no need to call `fmt::format` youself. 

For log messages, please adhere to noun-verb-object order, e.g., `VAST_DEBUG("pseudonymize operator skips field '{}' of unsupported type '{}'", ...)`.

Reply:
Fixed in https://github.com/tenzir/vast/pull/2719/commits/3dae63715e3a1f2e5444ff71519e7736cd3605bf